test/e2e_node/gcloud/gcloud.go

```diff
@@ -0,0 +1,174 @@
+}
+
+type gCloudClientImpl struct {
+	host string
+	zone string
+}
+
+type RunResult struct {
+	out []byte
+	err error
+	cmd string
+}
+
+type CmdHandle struct {
+	TearDown TearDown
+	Output   chan RunResult
+	LPort    string
+}
+
+func NewGCloudClient(host string, zone string) GCloudClient {
+	return &gCloudClientImpl{host, zone}
+}
+
+func (gc *gCloudClientImpl) Command(cmd string, args ...string) ([]byte, error) {
+	args = append([]string{"compute", "ssh", "--zone", gc.zone, gc.host, "--", cmd}, args...)
+	glog.Infof("Command %s", strings.Join(args, " "))
+	return exec.Command("gcloud", args...).CombinedOutput()
+}
+
+func (gc *gCloudClientImpl) TunnelCommand(sudo bool, lPort string, rPort string, cmd string, aa ...string) ([]byte, error) {
+	tunnelStr := fmt.Sprintf("-L %s:localhost:%s", lPort, rPort)
+	args := []string{"compute", "ssh", "--zone", gc.zone, "--ssh-flag", tunnelStr, gc.host, "--"}
+	if sudo {
+		args = append(args, "sudo")
+	}
+	args = append(args, cmd)
+	args = append(args, aa...)
+	glog.Infof("Command %s", strings.Join(args, " "))
+	return exec.Command("gcloud", args...).CombinedOutput()
+}
+
+func (gc *gCloudClientImpl) CopyToHost(from string, to string) ([]byte, error) {
+	rto := fmt.Sprintf("%s:%s", gc.host, to)
+	glog.Infof("Command gcloud compute copy-files --zone %s %s %s", gc.zone, from, rto)
+	return exec.Command("gcloud", "compute", "copy-files", "--zone", gc.zone, from, rto).CombinedOutput()
+}
+
+func (gc *gCloudClientImpl) CopyAndRunAsync(sudo bool, remotePort string, bin string, args ...string) *CmdHandle {
+	h := &CmdHandle{}
+	h.Output = make(chan RunResult)
+
+	rand.Seed(time.Now().UnixNano())
+
+	// Define where we will copy the temp binary
+	tDir := fmt.Sprintf("/tmp/gcloud-e2e-%d", rand.Int31())
+	_, f := filepath.Split(bin)
+	cmd := filepath.Join(tDir, f)
+	h.LPort = getLocalPort()
+
+	// Do the setup
+	go func() {
+		// Create the tmp directory
+		out, err := gc.Command("mkdir", "-p", tDir)
+		if err != nil {
+			glog.Errorf("mkdir failed %v", err)
+			h.Output <- RunResult{out, err, fmt.Sprintf("mkdir -p %s", tDir)}
+			return
+		}
+
+		// Copy the binary
+		out, err = gc.CopyToHost(bin, tDir)
+		if err != nil {
+			glog.Errorf("copy-files failed %v", err)
+			h.Output <- RunResult{out, err, fmt.Sprintf("copy-files %s %s", bin, tDir)}
+			return
+		}
+
+		// Start the process
+		out, err = gc.TunnelCommand(sudo, h.LPort, remotePort, cmd, args...)
+		if err != nil {
+			glog.Errorf("command failed %v", err)
+			h.Output <- RunResult{out, err, fmt.Sprintf("%s %s", cmd, strings.Join(args, " "))}
+			return
+		}
+	}()
+
+	h.TearDown = func() {
+		out, err := gc.Command("sudo", "pkill", f)
+		if err != nil {
+			h.Output <- RunResult{out, err, fmt.Sprintf("pkill %s", cmd)}
+			return
+		}
+		out, err = gc.Command("rm", "-rf", tDir)
+		if err != nil {
+			h.Output <- RunResult{out, err, fmt.Sprintf("rm -rf %s", tDir)}
+			return
+		}
+	}
+	return h
+}
+
+func (gc *gCloudClientImpl) CopyAndWaitTillHealthy(
```

Comment:
This seems to be a generic method. Can this be made independent of the cloud provider?